src/main/scala/is/hail/annotations/ReadRowsRDD.scala

```diff
@@ -1,23 +1,200 @@
 package is.hail.annotations
 
-import java.io.{DataInputStream, DataOutputStream}
+import java.io.{InputStream, OutputStream}
+import java.util
 
 import is.hail.expr._
 import is.hail.utils.{SerializableHadoopConfiguration, _}
-import net.jpountz.lz4.{LZ4BlockInputStream, LZ4Factory}
+import is.hail.variant.LZ4Utils
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.Row
 import org.apache.spark.{Partition, SparkContext, TaskContext}
 
-class Decoder(in: DataInputStream) {
+class ArrayInputStream(var a: Array[Byte], var end: Int) extends InputStream {
+  var off: Int = 0
 
-  def readByte(): Byte = {
-    val i = in.read()
-    assert(i != -1)
-    i.toByte
+  def this(a: Array[Byte]) = this(a, a.length)
+
+  def clear() {
+    off = 0
+  }
+
+  def read(): Int = {
+    if (off == end)
+      -1
+    else {
+      val c = a(off) & 0xff
```

Comment:
Is this how you convert a byte to an int? It doesn't make sense to me to apply a mask like this to an 8-bit value.

Reply:
Integral promotion rules convert Byte to Int (signed), so if you want the conversion unsigned, yes, this is the JVM idiom.  I'm sure the compiler handles it nicely.